File: 2020/app/src/main/scala/org/snowi/aoc2020/puzzle3/Puzzle3.scala
/* (C) stefan.devops@gmail.com 2021 */
package org.snowi.aoc2020.puzzle3

object Puzzle3 {

  val square = '.'
  val tree = '#'
  val slope11 = new Slope(1, 1)
  val slope31 = new Slope(3, 1)
  val slope51 = new Slope(5, 1)
  val slope71 = new Slope(7, 1)
  val slope12 = new Slope(1, 2)

  val slopes = Array(slope11, slope31, slope51, slope71, slope12)

  def main(args: Array[String]): Unit = {
//    logger.info("jjsjs")
    answer()
  }

  def answer(): Unit = {
    val input = readMap("puzzle3/input.txt")
    println()
    println("--------------- Puzzle 03 ---------------")
    println("Result puzzle 3 part 1: " + answer1(input, slope31))
    println("Result puzzle 3 part 2: " + answer2(input, slopes))
    println()
  }

  def answer1(map: List[String], slope: Slope): Int = {

    var line = 0
    var pos = 0
    var trees = 0
    while (line + slope.moveDown <= map.length - 1) {
      trees += move(map, line, pos, slope)
      line += slope.moveDown
      pos += slope.moveRight
    }
    println("checking slope " + slope + " --> " + trees)
    trees
  }

  def answer2(map: List[String], slopes: Array[Slope]): Int = {
    var result = 1
    for (slope <- slopes) {
      result *= answer1(map, slope)
    }
    result
  }

  def move(
      map: List[String],
      startLine: Int,
      startPos: Int,
      slope: Slope
  ): Int = {
  //  println("StartLine: " + startLine + "   startPos: " + startPos)
    val line = map(startLine + slope.moveDown)
    val newPosition = (startPos + slope.moveRight) % (line.length)
    val position = line.charAt(newPosition)
    if (position == tree) {
      1
    } else {
      0
    }

  }

  def readMap(filename: String): List[String] = {
    val bufferedSource = io.Source.fromResource(filename)
    val lines = (for (line <- bufferedSource.getLines()) yield line).toList
    bufferedSource.close
    lines
  }

}

File: 2020/app/src/main/java/org/snowi/aoc2020/puzzle3/Puzzle3.java
/* Copyright: (C) stefan.devops@gmail.com 2021, all rights reserved */
package org.snowi.aoc2020.puzzle3;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.snowi.aoc2020.Puzzle;

import java.util.List;

public class Puzzle3 extends Puzzle {

  private static final Logger logger = LoggerFactory.getLogger(Puzzle3.class);
  private static final char SQUARE = '.';
  private static final char TREE = '#';

  private final Slope slope11 = new Slope(1, 1);
  private final Slope slope31 = new Slope(1, 3);
  private final Slope slope51 = new Slope(1, 5);
  private final Slope slope71 = new Slope(1, 7);
  private final Slope slope12 = new Slope(2, 1);
  private final List<Slope> slopes = List.of(slope11, slope31, slope51, slope71, slope12);

  public Puzzle3(String inputFile) {
    super(inputFile);
  }

  @Override
  public String answer1(List<String> input) {
    return String.valueOf(slide(input, slope31));
  }

  private int slide(List<String> input, Slope slope) {
    int line = 0;
    int pos = 0;
    int trees = 0;
    while (line + slope.getMoveDown() <= input.size() - 1) {
      trees += move(input, line, pos, slope);
      line += slope.getMoveDown();
      pos += slope.getMoveRight();
    }
    logger.debug("checking slope " + slope + " --> " + trees);
    return trees;
  }

  @Override
  public String answer2(List<String> input) {
    int trees = 1;
    for (Slope slope : slopes) {
      trees *= slide(input, slope);
    }
    return String.valueOf(trees);
  }

  int move(List<String> map, int startLine, int startPos, Slope slope) {
    String line = map.get(startLine + slope.getMoveDown());
    int newPosition = (startPos + slope.getMoveRight()) % (line.length());
    char position = line.charAt(newPosition);
    return position == TREE ? 1 : 0;
  }
}
